extended repo LVBS kernel

ddstreet opened the pull request:
<!--
COMMENT BLOCKS WILL NOT BE INCLUDED IN THE PR.
Feel free to delete sections of the template which do not apply to your PR, or add additional details
-->

###### Merge Checklist  <!-- REQUIRED -->
<!-- You can set them now ([x]) or set them later using the Github UI -->
**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./LICENSES-AND-NOTICES/SPECS/data/licenses.json`, `./LICENSES-AND-NOTICES/SPECS/LICENSES-MAP.md`, `./LICENSES-AND-NOTICES/SPECS/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary <!-- REQUIRED -->
<!-- Quick explanation of the changes. -->
This is early work to build a kernel in the extended repo that includes LVBS support.

Note that the accumulated LVBS patches will be rolled into a dedicated LVBS kernel tarball, instead of carrying them as individual patches like this PR. Please ignore the actual content of the LVBS patches, as that content will be  reviewed/managed by a kernel git branch (and this PR shouldn't be merged until that is complete and this is updated to use the resulting LVBS kernel tarball); this PR is intended only for review of the spec file (and related) changes.


###### Change Log  <!-- REQUIRED -->
<!-- Detail the changes made here. -->
<!-- Please list any packages which will be affected by this change, if applicable. -->
<!-- Please list any CVES fixed by this change, if applicable. -->
Add spec file, etc. to enable building LVBS kernel in extended repo

###### Does this affect the toolchain?  <!-- REQUIRED -->
<!-- Any packages which are included in the toolchain should be carefully considered. Make sure the toolchain builds with these changes if so. -->
<!-- Update: manifests/package/toolchain_*.txt, pkggen_core_*.txt, update_manifests.sh -->
<!-- To validate: make clean; make workplan REBUILD_TOOLCHAIN=y DISABLE_UPSTREAM_REPOS=y CONFIG_FILE="" ... -->
NO

###### Test Methodology
<!-- How was this test validated? i.e. local build, pipeline build etc. -->
WIP, no testing/validation yet

reubeno commented:
Great to see the start of this, @ddstreet! I know we're mutually interested in seeing an LVBS-enabled kernel make its way into the extended repo.

Let us know when you're ready to start getting any early reviews/feedback on the PR. (I see it currently marked as WIP and draft.) I'm assuming @rlmenge, among others, will be keenly interested and have good suggestions.

rlmenge commented on SPECS-EXTENDED/kernel-lvbs/kernel-lvbs.spec:
```diff
@@ -0,0 +1,1925 @@
+
+    make mrproper
+
+    cp $CONFIG_FILE $KBUILD_OUTPUT/.config
```
Just checking but will this work since the `lvbs_normal` and `lvbs_secure` directories have not been created yet?

ddstreet replied:
yeah, `make mrproper` will create the `$KBUILD_OUTPUT` dir

rlmenge reviewed (commented):
Commits were super easy to follow, thank you! The logic makes sense to me but would want to validate in an rpmbuild environment to be sure all the KBUILD_OUTPUTs are operating as expected.

From what I gather with the spec is that it builds both secure and normal lvbs using the patches listed in the "kernel-lvbs.patches" and setting KBUILD_OUTPUT respectively. It then also builds Microsoft/skloader and Microsoft/initrd-sk.cpio.

A lot of the clean up work looks good to me such as making the config check its own function and reordering the manipulation of the config. Would just want again ensure the function is called as expected. 

Just need to also update License files and Manifests similar to `kernel-rt` and that will clear up the failing github checks

ddstreet commented:
> Great to see the start of this, @ddstreet! I know we're mutually interested in seeing an LVBS-enabled kernel make its way into the extended repo.
> 
> Let us know when you're ready to start getting any early reviews/feedback on the PR. (I see it currently marked as WIP and draft.) I'm assuming @rlmenge, among others, will be keenly interested and have good suggestions.

Thanks! It should be ready for initial review; I'll fix the failing checks and then remove the wip/draft. However, it's not yet ready for merging as it does still need to change over to use the dedicated tarball instead of separate patches.

ddstreet commented:
changes to the PR:
- updated PR with the new `lpg-innovate` kernel source tarball
- renamed kernel to `kernel-lpg-innovate` to match tarball tag/naming
- rebased on latest 3.0-dev
- restored local normal/secure kernel config files (not using in-tarball config files for rpm build)

ddstreet commented:
also please note the changes to `tdnf` and `dnf` packages:
- tdnf updated to add `kernel-lpg-innovate` package to existing `installonlypkgs` list in `tdnf.conf`
- -dnf updated to include a `installonlypkgs` entry in `dnf.conf` (did not previously have `installonlypkgs` defined for `dnf`)

ddstreet commented:
@rlmenge @reubeno @christopherco @tgopinath-microsoft this should be ready for review

ddstreet commented:
note that the `dracut.conf.d` that this adds will include *all* secure kernel binaries on the system into *all* initrds that are generated. I didn't see any easy way to adjust dracut config that would only include the correct version (if any) into each initrd, but i'm open to suggestions. My guess is it would probably require a new dracut module (i.e. `/usr/lib/dracut/modules.d/...`) which I didn't want to get into.

ddstreet commented:
> note that the `dracut.conf.d` that this adds will include _all_ secure kernel binaries on the system into _all_ initrds that are generated. I didn't see any easy way to adjust dracut config that would only include the correct version (if any) into each initrd, but i'm open to suggestions. My guess is it would probably require a new dracut module (i.e. `/usr/lib/dracut/modules.d/...`) which I didn't want to get into.

just to point it out, this can increase the size of each initrd significantly, since each secure kernel binary is (currently on my test system at least) around 15M. So if 3 `kernel-lpg-innovate` rpms are installed onto a system, all the initrd on the system  (including non-lpg-innovate kernel initrds) will be ~45M larger than normal. so it may be worth it to look into creating a dracut module (or some other approach) to include the secure kernel only where appropriate.

rlmenge commented on SPECS-EXTENDED/kernel-lpg-innovate/kernel-lpg-innovate.spec:
```diff
@@ -0,0 +1,1959 @@
+%description grub
+This package contains a grub config file to add required LPG-Innovate parameters to the kernel cmdline.
+
+%package -n     python3-perf
```
Can you add a short name to `bpftool` and `python3-perf` packages? Similar to https://github.com/microsoft/azurelinux/blob/3.0-dev/SPECS/kernel-64k/kernel-64k.spec#L6 

This avoids overwriting the kernel's version of these packages

ddstreet replied:
ok i added this.

I didn't add any `Conflicts:` (and/or `Provides:`) so the rpm file content of the packages will still conflict, of course.

rlmenge reviewed (approved):
Ran BB and saw success: https://dev.azure.com/mariner-org/mariner/_build/results?buildId=788032&view=results

Approve
